src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs

```diff
@@ -0,0 +1,294 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Sbom.Contracts;
+using Microsoft.Sbom.Contracts.Enums;
+using Microsoft.Sbom.Exceptions;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+
+namespace Microsoft.Sbom.Parser;
+
+/// <summary>
+/// Parses <see cref="SBOMFile"/> object from a 'files' array.
+/// </summary>
+internal ref struct SbomFileParser
+{
+    private readonly Stream stream;
+    private readonly SBOMFile sbomFile;
+
+    public SbomFileParser(Stream stream)
+    {
+        this.stream = stream ?? throw new ArgumentNullException(nameof(stream));
+
+        sbomFile = new ();
+    }
+
+    /// <summary>
+    /// Returns an iterator of <see cref="SBOMFile"/> objects by parsing the SPDX JSON files section.
+    /// </summary>
+    /// <param name="buffer">The buffer where the stream will be read.</param>
+    /// <param name="reader">The UTF8 reader used to parse the JSON.</param>
+    /// <param name="sbomFile">The object that eventually will be assigned.</param>
+    /// <returns></returns>
+    /// <exception cref="ParserError"></exception>
+    public long GetSbomFile(ref byte[] buffer, ref Utf8JsonReader reader, out SBOMFile sbomFile)
+    {
+        try
+        {
+            // If the end of the array is reached, return with null value to signal end of the array.
+            if (reader.TokenType == JsonTokenType.EndArray)
+            {
+                sbomFile = null;
+                return 0;
+            }
+
+            // Read the start { of this object.
+            ParserUtils.SkipNoneTokens(stream, ref buffer, ref reader);
+            ParserUtils.AssertTokenType(stream, ref reader, JsonTokenType.StartObject);
+
+            // Move to the first property name token.
+            ParserUtils.Read(stream, ref buffer, ref reader);
+            ParserUtils.AssertTokenType(stream, ref reader, JsonTokenType.PropertyName);
+
+            while (reader.TokenType != JsonTokenType.EndObject)
+            {
+                ParseProperty(ref reader, ref buffer);
+                
+                // Read the end } of this object or the next property name.
+                ParserUtils.Read(stream, ref buffer, ref reader);
+            }
+
+            // Validate the created object
+            ValidateSbomFile(this.sbomFile);
+
+            sbomFile = this.sbomFile;
+            return reader.BytesConsumed;
+        }
+        catch (EndOfStreamException)
+        {
+            sbomFile = null;
+            return 0;
+        }
+        catch (JsonException e)
+        {
+            sbomFile = null;
+            throw new ParserError($"Error while parsing JSON, addtional details: ${e.Message}");
```

Comment:
>ParserError [](http://example.com/codeflow?start=22&length=11)

Maybe add the JsonException and InnerException?

Reply:
Fixed